sled-agent/src/hardware/illumos/disk.rs

```diff
@@ -0,0 +1,407 @@
+}
+
+/// Parses, validates, and ensures the partition layout within a disk.
+///
+/// Returns a Vec of partitions on success. The index of the Vec is guaranteed
+/// to also be the index of the partition.
+pub fn ensure_partition_layout(
+    log: &Logger,
+    paths: &DiskPaths,
+    variant: DiskVariant,
+) -> Result<Vec<Partition>, DiskError> {
+    internal_ensure_partition_layout::<libefi_illumos::Gpt>(log, paths, variant)
+}
+
+// Same as the [ensure_partition_layout], but with generic parameters
+// for access to external resources.
+fn internal_ensure_partition_layout<GPT: gpt::LibEFIGpt>(
+    log: &Logger,
+    paths: &DiskPaths,
+    variant: DiskVariant,
+) -> Result<Vec<Partition>, DiskError> {
+    // Open the "Whole Disk" as a raw device to be parsed by the
+    // libefi-illumos library. This lets us peek at the GPT before
+    // making too many assumptions about it.
+    let raw = true;
+    let path = paths.whole_disk(raw);
+
+    let gpt = match GPT::read(&path) {
+        Ok(gpt) => {
+            // This should be the common steady-state case
+            info!(
+                log,
+                "Disk at {} already has a GPT",
+                paths.devfs_path.display()
+            );
+            gpt
+        }
+        Err(libefi_illumos::Error::LabelNotFound) => {
+            // Fresh U.2 disks are an example of devices where "we don't expect
+            // a GPT to exist".
+            info!(
+                log,
+                "Disk at {} does not have a GPT",
+                paths.devfs_path.display()
+            );
+
+            // For ZFS-implementation-specific reasons, Zpool create can only
+            // act on devices under the "/dev" hierarchy, rather than the device
+            // path which exists in the "/devices" directory.
+            let dev_path = if let Some(dev_path) = &paths.dev_path {
+                dev_path
+            } else {
+                return Err(DiskError::CannotFormatMissingDevPath { path });
+            };
+            match variant {
+                DiskVariant::U2 => {
+                    info!(
+                        log,
+                        "Formatting zpool on disk {}",
+                        paths.devfs_path.display()
+                    );
+                    // If a zpool does not already exist, create one.
+                    let zpool_name = ZpoolName::new(Uuid::new_v4());
+                    Zpool::create(zpool_name, dev_path)?;
+                    return Ok(vec![Partition::ZfsPool]);
+                }
+                DiskVariant::M2 => {
+                    // TODO: If we see a completely empty M.2, should we create
+                    // the expected partitions? Or would it be wiser to infer
+                    // that this indicates an unexpected error conditions that
+                    // needs mitigation?
+                    todo!("Provisioning M.2 devices not yet supported");
```

Comment:
How long do we expect this `todo!()` to be around? Would it be safer to make it an error instead?

Reply:
Yeah, I'll throw an error instead. Updated.